### DISCUSSION TITLE
BMD Drivers 14.3 Add support for CVPixelBuffer for improved performance.

### DISCUSSION THREAD

**Discussion by alatteri** [CESNET/UltraGrid#425] [General]
In the latest BMD Drivers 14.3, "Add support for CVPixelBuffer for improved performance."
Will this have any latency improvement within UG?

**Comment by MartinPulec** [comment]
I've looked into it and it should be related to some mac-specific extension.

In particular, the mac cvPixelBuffer can be directly passed as the DeckLink output frame. I don't think that UltraGrid can benefit from that now directly. The point is that you'd perhaps need to get the buffer from another macOS API, rather than creating own. I believe that the advantage is that the DeckLink frame retains ownership so that the data actually doesn't need to be copied. In case of UG, the data to the output pixel is copied anyways, so it may not have any advantage.

But I am not 100% sure, there can be concept like [CUDA pinned memory](https://developer.nvidia.com/blog/how-optimize-data-transfers-cuda-cc/#pinned_host_memory), which speeds-up actual PCIe transmissions.

**Reply by alatteri** [reply]
Ok thank.... I use Linux anyway.  Didn't realize this was a MacOS only thing.

**Reply by MartinPulec** [reply]
sure, no problem. I am glad that you notified me about the new release - with help of this release I've perhaps understood SDK/driver version compatibility (14.3 actually breaks ABI, but the last ABI bump was 11.6 so I believe that UG compiled with  SDK14.2 could be compatible with drivers as old as 11.6.
